### DISCUSSION TITLE
Making objects easier to create

### DISCUSSION THREAD

**Discussion by nathanvanthof** [Dlubal-Software/RFEM_Python_Client#255] [Ideas]
Hi everyone,

I've recently followed the RFEM python training. One thing that struck me was that you have to quite meticulously follow which node/member has what number. In my opinion this will make code difficult to maintain further down the road when all the node numbers slowly become "magic" (similarly to how regex becomes kind of gibberish after looking away for 5 minutes).

I would like to be able to skip this. For instance, by creating a member with only passing 2 Node and 1 Section object. I've started off creating a new module which basically adds a very simple interfacing layer. But there's a lot of duplicate docstrings; I'm thinking it might be nicer to add the functionality straight to the existing objects, but I'm not sure if that would break any backwards compatibility issues.

On the other hand, it specifically states "There will be broken objects or adjustments affecting backward compatibility." so that's why I think it could be better to implement it there?

PS. I've done a fair bit of programming, but this would be the first time participating in open source. I'm pretty unaware of any conventions, so feel free to tell me if I'm doing this wrong

**Comment by dogukankaratas** [comment]
Hi,

Thanks for the feedback. When developing the API, we'd like to stick to the logic of RFEM 6 itself. In my opinion, if we start to create these shortcuts, our library will be more and more complex and will be hard to understand for structural engineers.

But of course, there is no limitation for your creativitiy 😄 For example a simple function like below could work. 

```python
def easyMember(firstNode, secondNode, material, section):

    # nodes
    Node(1, firstNode[0], firstNode[1], firstNode[2])
    Node(2, secondNode[0], secondNode[1], secondNode[2])

    # material
    Material(1, material)

    # section
    Section(1, section, 1)

    # member
    Member(1, 1, 2, 0, 1, 1)
```

If you want to contribute our library, it'd be good to go over the open issues and pull requests.

**Reply by nathanvanthof** [reply]
I was thinking more something along the lines of this:
https://github.com/nathanvanthof/RFEM_Python_Client/tree/main/RFEM/EzObjects

But maybe that would be a better for a separate repository then


I'll have a look at the open issues

**Reply by heetrojivadiya** [reply]
> I was thinking more something along the lines of this: https://github.com/nathanvanthof/RFEM_Python_Client/tree/main/RFEM/EzObjects
> 
> But maybe that would be a better for a separate repository then
> 
> I'll have a look at the open issues

Hey @nathanvanthof ,

I have checked your codes for EzObjects and they look really cool. 

I have just one question: For example if I create 4 nodes by using ezNode like node(0,0,0), node(1,0,0), node(1,0,-1) and node(0,0,-1). Now,  I want to create member on node(0,0,0) and node(1,0,-1) using ezMember like member(start_section: Section, line: Line = None, node1: Node = None, node2: Node = None, end_section: Section = None, rotation_angle: float = 0.0, start_member_hinge: memberHinge = None, end_member_hinge: memberHinge = None) then what number should I have to pass as node1 and node2 in member()?

**Reply by nathanvanthof** [reply]
You wouldn't pass a number, but a node object. so:

    start_beam = node(0,0,0)
    end_beam = node(1,0,-1)
    another_node = node(1,0,0)
    last_node = node(0,0,-1)

    my_beam = beam(node1=start_beam, node_2=end_beam, ...)
    
Using objects rather than "magic numbers" is more pythonic imo

**Comment by OndraMichal** [comment]
Hi @nathanvanthof ,
we are not using "magic numbers". We are using identification number (id) which percisely defines object which, in case of nodes, has more than 35 parameters. Your case represents very specific use case where you want to use small number of objects defined only by 3 parameters. 

What would it look like when you want to use 1000 nodes each defined by 30 parameters. Would you rather use name like start_node_beam_3rd_row_5th_column_brace? I think not. Its easier to use id. Also, 3 coordinates quickly become magic numbers when you try to define more than 100 nodes.

As @dogukankaratas said. The library would quickly become clouded with alternate ways to set up each object. That said, you can still define your own functions. It is just not applicable to wide range of use cases.

**Reply by nathanvanthof** [reply]
@OndraMichal I completely understand. I'm sorry, I didn't mean to suggest RFEM did not clearly define this, I was just a bit confused initially. I imagine with time I'll get used to this notation and it'll become like 2nd nature for me :)

Just to answer your question:

I was thinking something like:

     start = "start"
     peak = "peak"
     end = "end"

     my_nodes = []
     for row in range(10):     
         row = {}
         for node in [start, peak, end]:
             row [node] = node(foo bar)
         my_nodes.append(row)

And then to call nodes based on row and position (`start`, `peak` or `end`).
